pandas/core/algorithms.py

```diff
@@ -415,33 +417,47 @@ def isin(comps, values):
     comps = com._values_from_object(comps)
 
     comps, dtype, _ = _ensure_data(comps)
-    values, _, _ = _ensure_data(values, dtype=dtype)
+
+    is_time_like = lambda x: (is_datetime_or_timedelta_dtype(x)
+                              or isinstance(x, Timestamp))
+
+    if is_time_like(dtype):
+        values, _, _ = _ensure_data(values, dtype=dtype)
+    else:
+        values, _, _ = _ensure_data(values)
 
     # faster for larger cases to use np.in1d
-    f = lambda x, y: htable.ismember_object(x, values)
+    f = lambda x, y: htable.ismember_object(x.astype(object), y.astype(object))
+
+    comps_types = set(type(v) for v in comps)
+    values_types = set(type(v) for v in values)
```

Comment:
you are going to great lengths to circument the structure below. This needs to be an extension of the if/then. I don't want to see ``int_flg`` or anything like that. This is also not performant. The reason we check dtypes in this way is to avoid conversions and materialization.

Reply:
Thanks - done
ASV results on algorithms after 82386c3 are similar to the most recent one 5416711 
```
       before           after         ratio
     [537b65cb]       [82386c31]
-     2.36±0.02ms      1.92±0.02ms     0.82  algorithms.Hashing.time_series_timedeltas

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```